### ISSUE TITLE
Provide Trilinos-global CMake options to make it easier for apps to disable float & complex Scalar types

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#362]
**Next Action Status:**

Make the changes in the `better-ci-build-482` branch as part of #482 ...

**CC:** @trilinos/framework, @trilinos/tpetra

**Blocked By:** #158

**Blocking:** #482

**Description:**

By default, in release and development mode, Trilinos provides support for float and complex (`std::complex<float>` and `std::complex<double>`) scalar types.  When  ETI is used, this massively increases the compile and link times for Trilinos.  Most Trilihnos users don't even need float or complex scalar types.

Currently, to disable float and complex , the user has to set a bunch of separate Teuchos and Tpetra options.  Very few users know what options these are so they get long build times by default and no idea how to disable them.

This story is to change the default for Trilinos to not build float or complex types by default.  This story will also provide top level build options `Trilinos_ENABLE_FLOAT` and `Trilinos_ENABLE_COMPLEX` (and  `Trilinos_ENABLE_COMPLEX_FLOAT` and `Trilinos_ENABLE_COMPLEX_DOUBLE` with obvious defaults based on `Trilinos_ENABLE_FLOAT` and `Trilinos_ENABLE_COMPLEX`) that all Trilinos packages that currently have their own variables will respond to. 

**Definition of Done:**
1. By default Trilinos builds without support for (or ETI instantiations for) float or complex scalar types.
2. Users can turn on support for float and/or complex types by setting `-DTrilinos_ENABLE_FLOAT=ON` and/or `-DTrilinos_ENABLE_COMPLEX=ON`, respectively (or something similar) ... 

**Tasks:**
1. Made a decision if support and explicit instantations for float and/or complex should be enabled by default or not. [DONE]
2. Add Trilinos repository options `Trilinos_ENABLE_FLOAT`, `Trilinos_ENABLE_COMPLEX`, `Trilinos_ENABLE_COMPLEX_FLOAT` and `Trilinos_ENABLE_COMPLEX_DOUBLE` and make Teuchos, Tpetra, Thyra, and other packages set their defaults based on this. (These should be Trilinos repository level options so that they can be set in meta projects like CASL VERA, MPACT, SCALE, as well)
3. Explicitly turn on `Trilinos_ENABLE_FLOAT`, `Trilinos_ENABLE_COMPLEX`, `Trilinos_ENABLE_COMPLEX_FLOAT` and `Trilinos_ENABLE_COMPLEX_DOUBLE` in a few automated Trilinos builds.

**Comment by mhoemmen** [comment]
In Tpetra, `Scalar = std::complex<double>` and `std::complex<float>` are only enabled by default in a debug build.  `Scalar = float` is always disabled by default.

Imagine that a supercomputer software vendor wants to support Trilinos.  Do you think they will ask us how to enable complex Scalar types?  Of course not.  They will configure with the minimal settings to make it build, then install the thing.  Users then get `Scalar = double` and that's it.

Applications can and do disable types that they don't need.  Also, we could always make Tpetra's ETI work like the "ETI" in KokkosKernels, where any Scalar type works, but some types get pre-compiled kernels by default.  The new `extern template` feature of C++11 might help make this easier.

I like your second feature request.  A while back, I was trying to add `Scalar = __float128` support to Tpetra.  I found out the hard way that Thyra uses a separate set of ETI macros managed through Teuchos.  It would make sense for Thyra and Tpetra to share the same ETI system, even if one doesn't necessarily support all the type combinations that the other supports.

**Comment by bartlettroscoe** [comment]
> In Tpetra, `Scalar = std::complex<double>` and `std::complex<float>` are only enabled by default in a debug build.  `Scalar = float` is always disabled by default.

I am not sure that we want to change what types are enabled by default based on a debug build or not.  I think that would confuse users (I would be surprised and confused by this).

> Imagine that a supercomputer software vendor wants to support Trilinos.  Do you think they will ask us how to enable complex Scalar types?  Of course not.  They will configure with the minimal settings to make it build, then install the thing.  Users then get `Scalar = double` and that's it.

I disagree.  We make it clear in the Trilinos/INSTALL document that by default you only get support for `double` and if they want support for `float` and `std:complex`, then they need to add `-DTrilinos_ENBALE_FLOAT=ON` and `-DTrilinos_ENBALE_COMPLEX=ON`.  I think that is the way to go.  Otherwise, potential new Trilinos users will get turned off immediately when they try to build Trilinos the first time and it takes hours and hours to build a bunch of instantations that 99% of them will never use.

If super computing centers are installing Trilinos in a vacuum (i.e. not talking to Trilinos developers or Trilinos users), then people end up ignoring the default installed version of Trilnos anyway.  How many people do you know who use the default installed version of Trilinos on these machines?  Currently, if they don't enable enough TPLs, then people ignore the default installed version of Trilinos.  Hopefully with Spack that will change (and we can make sure that Trilinos is configured appropriately with Spack which will include if float and complex types are enabled or not).

> Applications can and do disable types that they don't need.

Close application customers are not a problem either way. We just tell them how to configure and they do it.  The issue is new potential customers that are getting ridiculously long build times by default.  That is the issue.

> Also, we could always make Tpetra's ETI work like the "ETI" in KokkosKernels, where any Scalar type works, but some types get pre-compiled kernels by default.  The new `extern template` feature of C++11 might help make this easier.

Not sure what `extern template` buys you but it is worth looking into.

> I like your second feature request.  A while back, I was trying to add `Scalar = __float128` support to Tpetra.  I found out the hard way that Thyra uses a separate set of ETI macros managed through Teuchos.  It would make sense for Thyra and Tpetra to share the same ETI system, even if one doesn't necessarily support all the type combinations that the other supports.

I don't have any major objections to this but Tpetra is just an optional dependency for Thyra so it might take some work to get this to behave correctly.  I think we would really rather have the selection of Scalar (and Ordinal) types specified by default at the global level and then the various Trilinos packages should use those, as described above for float and complex types.

We should discuss this at the Trilinos Leaders Meeting today.

**Comment by mhoemmen** [comment]
I'm renaming this, because not everyone agrees about _default_ behavior.  We agree that such options should be shared among Trilinos packages (in particular between Thyra, Tpetra, and their downstream packages).  We also agree that it should be easy for installers and applications to change the settings.

**Comment by bartlettroscoe** [comment]
> because not everyone agrees about default behavior.

Everyone does not have to agree, but we still need to define default behavior.  We will discuss this at the next Trilinos leaders meeting but I will argue that the massive build times for Trilinos should make the default to exclude float and complex types in the default configure of Trilinos.

**Comment by bartlettroscoe** [comment]
I added blocking relationship with #158 since I don't think we can test such a change without building almost all of Trilinos (which is very hard to do right now).

**Comment by mhoemmen** [comment]
Should we let users specify a list of enabled e.g., Scalar types, rather than saying Trilinos_ENABLE_FLOAT etc.?  This could get confusing, e.g., if users want to enable Scalar=int (does Trilinos_ENABLE_INT mean Scalar, LocalOrdinal, or GlobalOrdinal?).

**Comment by bartlettroscoe** [comment]
At the [Trilinos Leaders meeting today](https://docs.google.com/document/d/1HYjscqKKQUOzW99soaeE_w3AtPD5GH8yHfZf2sKE0e0/), we discussed this if we should enabled float and complex by default or not.  The consensus decision was to **disable float and complex by default** but make it easy and obvious for how to enable them.

Now the question is how to enable them?

> Should we let users specify a list of enabled e.g., Scalar types, rather than saying Trilinos_ENABLE_FLOAT etc.? This could get confusing, e.g., if users want to enable Scalar=int (does Trilinos_ENABLE_INT mean Scalar, LocalOrdinal, or GlobalOrdinal?).

That is not a bad idea.  I might suggest:

```
-DTrilinos_ENABLE_SCALAR_TYPES="float, complex, complex<float>, complex<double>, ..."
```
(the type `double` would be hard coded and could not be disabled.  Or we could supply a separate option to disable the default `double` support and instantations.)

So above, if the user specified `float,complex`, then they would actually get support (and instantations with ETI enabled) for `complex<float>` and `complex<double>`.  If they did not want, for example, `complex<float>`, then they would specify `float,complex<double>`.  Sound good?

Then we could do something similar to for GlobalOrdinal andLocalOrdinal:

```
-DTrilinos_ENABLE_LOCAL_INT_TYPES="int, long int, long long int, ..."
-DTrilinos_ENABLE_GLOBAL_INT_TYPES="long long int, long int, int, ..."
```

These types of CMake arguments would require more CMake work to parse this (i.e.. `long long` and `long long int` are the same) but it would allow for better extendability.

Comments, opinions?

**Comment by wfspotz** [comment]
I think the lists approach is the right way to go. I have a couple questions:

1. Does `complex<double>` work without the `std` namespace? If so, why? What is the rule for type specification? Is `using std` assumed when parsing the list?
2. Are there any issues with user-defined types (presumably for `SCALAR`)?

**Comment by mhoemmen** [comment]
> Does `complex<double>` work without the `std` namespace? If so, why? What is the rule for type specification? Is `using std` assumed when parsing the list?

I think @bartlettroscoe was just abbreviating -- it would make sense for the list to be parsed as if it occurred in the global namespace with no `using` declarations.

**Comment by bartlettroscoe** [comment]
I think we could allow the CMake parsing to be smart enough to allow `std::complex<double>` or just `complex<double>`.  We just need to document this.

The bigger question is if this will also support addon user-defined types?  That is a another type of requirement that I would rather address in another Story ticket.

I would guess that given the current scope, this should be about a days work (i.e. 8 dev hours) including automated tests, full manual testing on Trilinos, documentation, etc.  But this is the easiest thing that we can do that will result in a dramatic reduction in default build times (which we should time).  But before we start this, we should do #158 to make this easy to test.

**Comment by mhoemmen** [comment]
> The bigger question is if this will also support addon user-defined types? That is a another type of requirement that I would rather address in another Story ticket.

I agree that we don't have to go there for now.  Any type that requires including a header file would _not_ work without further modifications.  Built-in but nonstandard types that don't require including a header file (or already have it included) should work fine.  This includes GCC's `__float128` (as long as users link with libquadmath).

**Comment by mhoemmen** [comment]
> I think we could allow the CMake parsing to be smart enough to allow std::complex<double> or just complex<double>. We just need to document this.

Don't be smart :-)

**Comment by mhoemmen** [comment]
> But this is the easiest thing that we can do that will result in a dramatic reduction in default build times (which we should time).

A lot of the build time stuff is not ETI-related, it's that we instantiate too much code.  To me, this issue is about simplifying Tpetra - Thyra - Stratimikos integration, and improving the user experience.

**Comment by bartlettroscoe** [comment]
> Let's make users spell this out for us. How do we know the difference between std::complex and Kokkos::complex otherwise?

I think for the purposes of a global option to enable complex support, it should be just `complex` and that should enable or disable std::complex and Kokkos::complex.  If package's want package-specific finer-grained enables/disables, then that is fine.

I am actually about to implement this soon since this will be important to create a more efficient pre-push testing process.

**Comment by bartlettroscoe** [comment]
Regardless what we use as the final user interface, we are going to want simple CMake variables like `${PROJECT_NAME}_ENABLE_FLOAT` and  `${PROJECT_NAME}_ENABLE_COMPLEX` to make it easy to set the defaults for the vars:

* Teuchos_ENABLE_FLOAT
* Teuchos_ENABLE_COMPLEX
* Sacado_ENABLE_COMPLEX
* Thyra_ENABLE_COMPLEX
* Tpetra_INST_COMPLEX_DOUBLE
* Tpetra_INST_COMPLEX_FLOAT
* Anasazi_ENABLE_COMPLEX

That should be fast and easy to implement for now.

**Comment by crtrott** [comment]
Just a comment to Kokkos vs std complex: std:: complex doesn't work in parallel environments well, since it a) doesn't exist on GPUs, and b) misses volatile overloads of its operators which prevents things like Kokkos reductions to work. What we did with Tpetra is that you request std::complex, but internally will get Kokkos::complex. The two types are interoperable (i.e. assignable and promotable to each other).

**Comment by bartlettroscoe** [comment]
Responding to @crtrott in [above comment](https://github.com/trilinos/Trilinos/issues/362#issuecomment-228772012):

> Just a comment to Kokkos vs std complex: std:: complex doesn't work in parallel environments well, since it a) doesn't exist on GPUs, and b) misses volatile overloads of its operators which prevents things like Kokkos reductions to work. What we did with Tpetra is that you request std::complex, but internally will get Kokkos::complex. The two types are interoperable (i.e. assignable and promotable to each other).

I think that means that it makes no sens to say that you are supporting "Kokkos::complex" and "std::complex".   It will just be "complex" and the Kokkos::complex will be used where it needs to be and std::complex can be used else where.  I that a fair assessment?

**Comment by crtrott** [comment]
Yeah that is pretty much what I think about this. Kokkos complex is not meant as an alternative to std::complex, it is meant to supplement it since std::complex has deficiencies which would prevent its usage throughout the threaded Trilinos stack.

**Comment by bartlettroscoe** [comment]
> Yeah that is pretty much what I think about this. Kokkos complex is not meant as an alternative to std::complex, it is meant to supplement it since std::complex has deficiencies which would prevent its usage throughout the threaded Trilinos stack. 

Right, so that argues that my recommendation in the [above comment](https://github.com/trilinos/Trilinos/issues/362#issuecomment-221363070) to use:

```
-DTrilinos_ENABLE_SCALAR_TYPES="float, complex, complex<float>, complex<double>, ..."
```

It makes no sense to differentiate "std::complex" from "complex" because Trilinos will pick the right complex impl for the given situation.  The user would interact with Kokkos::complex or std::complex or a C-struct or Fortran equivalent depending on their situation.  It is just "complex".

**Comment by crtrott** [comment]
Agreed.

**Comment by bartlettroscoe** [comment]
I just realized that we really need to get this Story done before we can deploy the new CI build that has float and complex disabled (see #482).  If we don't, then any build failure with complex or float will take down **all** of the Nightly builds.  So to support complex, we need to run an additional post-push CI server that turns these on (or at least the complex types that users really use).

Therefore, I will implement this on the branch `better-ci-build-482` and test this.  After the new CI build is deployed, we can update a few of the other automated Trilinos builds to turn these back on.

**Comment by bartlettroscoe** [comment]
The Trilinos JIRA isssue:

* https://sems.sandia.gov/jira/projects/TRIL/issues/TRIL-103

will be used to collect responses from internal Trilinos customers for if they use float and complex types.

I will now ask external Trilinos users about who uses float and complex types.

**Comment by bartlettroscoe** [comment]
I sent out the below email to Trilinos users to inquire about usage of float and complex types.

----

From: Bartlett, Roscoe A 
Sent: Thursday, November 17, 2016 8:13 AM
To: <trilinos-users@trilinos.org>
Subject: Usage of float and complex scalars with templated Trilinos?

Hello Trilinos users,

As part of trying to make Trilinos easier and less painful for average users to build (and make our CI build faster), we are going to turn off float and complex support by default.  But we will provide top-level configure options to make it easy to turn them on.  See:

     https://github.com/trilinos/Trilinos/issues/362 

But float and complex still needs to be tested to support customers that use these types.  We are just not sure where we should target our testing for these (since enabling these types massively increases the build and runtimes for many Trilinos libraries and tests).

Therefore, do you know of Trilinos customers that are relying on float and/or complex scalar types with Trilinos?  And if they are using std::complex<T>, is it only std::complex<double> or is it also std::complex<float>?

You can reply directly to me or you can directly add a comment to:

     https://github.com/trilinos/Trilinos/issues/362 

Thanks,

-Ross

**Comment by bartlettroscoe** [comment]
It seems that we have at least on important internal Sandia customer of Trilinos that uses `std::complex<double>`.  See:

* https://sems.sandia.gov/jira/browse/TRIL-103?focusedCommentId=13334&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13334 

Therefore, because of the importance of that customer, we should provide another optional `--default-build` as part of checkin-test.py that turns on `std::complex<double>` called something like `MPI_RELEASE_DEBUG_SHARED_COMPLEX_DOUBLE` and set up another post-push CI server that tests with `std::complex<double>` enabled (i.e. `-DTrilinos_ENABLE_COMPLEX_DOUBLE=ON`).  The checkin-test-sems.sh script would default to the one build without complex with `--default-builds=MPI_DEBUG_RELEASE_SHARED`.  We can then recommend developers who are messing with code that could break the complex build to also run this build on pre-push testing to be safe.  But most developers could likely skip this and we could rely on the post-push CI build `MPI_RELEASE_DEBUG_SHARED_COMPLEX_DOUBLE` to catch what falls through the cracks before the Nightly builds fire off.

**Comment by bartlettroscoe** [comment]
I have added support for these global variables and did some fairly detailed manual testing that is detailed below.

I pushed the commit to the `better-ci-build-482` branch so that I can use this as part fo that branch.

I am putting this in review.

----

**Detailed Notes:**

Working on the branch `beter-ci-build-482`.

**1) Create new options Trilinos_ENABLE_FLOAT, Trilinos_ENABLE_COMPLEX, Trilinos_ENABLE_COMPLEX_FLOAT and Trilinos_ENABLE_COMPLEX_DOUBLE in Trilinos/cmake/CallbackSetupExtraOptions.cmake.** [Done]

**2) Use these new options throughout all of the places they need to go and test configures of Trilinos (with them on and off):** [Done]

**3) Remove default disables in Trilinos/cmake/CallbackSetupExtraOptions.cmake:** [Done]

**4) Do detailed testing with differnet scalar types enabled and disabled:**

**4.1) Default configure (float and complex off):**

Generate the configure file with:

```
$ ./checkin-test-sems.sh --enable-all-packages=on --allow-no-pull --configure
```

Inspecting the CMake configure output:

```
Processing explicit instantiation support for enabled packages ...

Processing ETI support: Ifpack2
-- Ifpack2: Processing ETI / test support
-- Enabled Scalar types:        double
-- Enabled LocalOrdinal types:  int
-- Enabled GlobalOrdinal types: int|long long
-- Enabled Node types:          Kokkos::Compat::KokkosSerialWrapperNode
-- User/Downstream ETI set: 
-- Excluded type combinations: 
-- Set of enabled types, before exclusions: S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long}
Processing ETI support: TpetraCore
-- TpetraCore: Processing ETI / test support
-- Enabled Scalar types:        int|long long|double
-- Enabled LocalOrdinal types:  int
-- Enabled GlobalOrdinal types: int|long long
-- Enabled Node types:          Kokkos::Compat::KokkosSerialWrapperNode
-- Excluded type combinations: :
-- Set of enabled types, before exclusions: S={int} LO={int} GO={int} N={Kokkos::Compat::KokkosSerialWrapperNode};S={int} LO={int} GO={long long} N={Kokkos::Compat::KokkosSerialWrapperNode};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={int} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={int} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={long long} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={long long} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long}
Processing ETI support: TpetraKernels
-- TpetraKernels: Processing ETI / test support
-- Enabled Scalar types:       int|long long|double
-- Enabled LocalOrdinal types: int
-- Enabled Device types:       Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>
-- Set of enabled types, before exclusions: S={int} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={long long} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={double} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>}
```

Now inspecting the CMakeCache.txt file:

```
$ grep _FLOAT CMakeCache.txt  | grep -vi complex
Teuchos_ENABLE_FLOAT:BOOL=FALSE
Thyra_ENABLE_FLOAT:BOOL=OFF
Tpetra_INST_FLOAT:BOOL=OFF
Tpetra_INST_FLOAT128:BOOL=OFF
Trilinos_ENABLE_FLOAT:BOOL=FALSE
HAVE_MUELU_INST_FLOAT_INT_INT:INTERNAL=OFF
HAVE_TEUCHOS_FLOAT:INTERNAL=OFF
HAVE_THYRA_FLOAT:INTERNAL=OFF
HAVE_TPETRA_FLOAT:INTERNAL=OFF
HAVE_TPETRA_FLOAT128:INTERNAL=OFF
HAVE_TPETRA_INST_FLOAT:INTERNAL=OFF
HAVE_TPETRA_INST_FLOAT128:INTERNAL=OFF
MueLu_INST_FLOAT_INT_INT:INTERNAL=OFF
Tpetra_INST_FLOAT_DEFAULT:INTERNAL=OFF

$ grep _COMPLEX CMakeCache.txt
Anasazi_ENABLE_COMPLEX:BOOL=OFF
Sacado_ENABLE_COMPLEX:BOOL=FALSE
Teuchos_ENABLE_COMPLEX:BOOL=FALSE
Thyra_ENABLE_COMPLEX:BOOL=OFF
Tpetra_INST_COMPLEX_DOUBLE:BOOL=OFF
Tpetra_INST_COMPLEX_FLOAT:BOOL=OFF
Trilinos_ENABLE_COMPLEX:BOOL=FALSE
Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=FALSE
Trilinos_ENABLE_COMPLEX_FLOAT:BOOL=FALSE
HAVE_ANASAZI_COMPLEX:INTERNAL=OFF
HAVE_KOKKOSTSQR_COMPLEX:INTERNAL=OFF
HAVE_ML_COMPLEX_MAXWELL:INTERNAL=OFF
HAVE_MUELU_INST_COMPLEX_INT_INT:INTERNAL=OFF
HAVE_SACADO_COMPLEX:INTERNAL=OFF
HAVE_TEUCHOS_COMPLEX:INTERNAL=OFF
HAVE_THYRA_COMPLEX:INTERNAL=OFF
HAVE_TPETRA_COMPLEX_DOUBLE:INTERNAL=OFF
HAVE_TPETRA_COMPLEX_FLOAT:INTERNAL=OFF
HAVE_TPETRA_INST_COMPLEX_DOUBLE:INTERNAL=OFF
HAVE_TPETRA_INST_COMPLEX_FLOAT:INTERNAL=OFF
MueLu_INST_COMPLEX_INT_INT:INTERNAL=OFF
Tpetra_INST_COMPLEX_DOUBLE_DEFAULT:INTERNAL=OFF
Tpetra_INST_COMPLEX_FLOAT_DEFAULT:INTERNAL=OFF
```

That all looks correct.

Now to try the build and tests:

```
???
```

**4.2) float and complex on:**

In the directory MPI_RELEASE_DEBUG_SHARED created by the checkin-test-sems.sh script, run:

```
$ time ./do-configure -DTrilinos_ENABLE_FLOAT=ON -DTrilinos_ENABLE_COMPLEX=ON &> configure.out
```

Looking at the cmake configure output:

```
Processing explicit instantiation support for enabled packages ...

Processing ETI support: Ifpack2
-- Ifpack2: Processing ETI / test support
-- Enabled Scalar types:        float|double|std::complex<float>|std::complex<double>
-- Enabled LocalOrdinal types:  int
-- Enabled GlobalOrdinal types: int|long long
-- Enabled Node types:          Kokkos::Compat::KokkosSerialWrapperNode
-- User/Downstream ETI set: 
-- Excluded type combinations: 
-- Set of enabled types, before exclusions: S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<float>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<float>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long}
Processing ETI support: TpetraCore
-- TpetraCore: Processing ETI / test support
-- Enabled Scalar types:        int|long long|float|double|std::complex<float>|std::complex<double>
-- Enabled LocalOrdinal types:  int
-- Enabled GlobalOrdinal types: int|long long
-- Enabled Node types:          Kokkos::Compat::KokkosSerialWrapperNode
-- Excluded type combinations: :
-- Set of enabled types, before exclusions: S={int} LO={int} GO={int} N={Kokkos::Compat::KokkosSerialWrapperNode};S={int} LO={int} GO={long long} N={Kokkos::Compat::KokkosSerialWrapperNode};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<float>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<float>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={int} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={int} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={long long} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={long long} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<float>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<float>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long}
Processing ETI support: TpetraKernels
-- TpetraKernels: Processing ETI / test support
-- Enabled Scalar types:       int|long long|float|double|Kokkos::complex<float>|Kokkos::complex<double>
-- Enabled LocalOrdinal types: int
-- Enabled Device types:       Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>
-- Set of enabled types, before exclusions: S={int} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={long long} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={float} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={double} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={Kokkos::complex<float>} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={Kokkos::complex<double>} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>}
```

and the CMakeCache.txt file:

```
$ grep _FLOAT CMakeCache.txt  | grep -vi complex
Teuchos_ENABLE_FLOAT:BOOL=TRUE
Thyra_ENABLE_FLOAT:BOOL=ON
Tpetra_INST_FLOAT:BOOL=ON
Tpetra_INST_FLOAT128:BOOL=OFF
Trilinos_ENABLE_FLOAT:BOOL=ON
HAVE_MUELU_INST_FLOAT_INT_INT:INTERNAL=ON
HAVE_TEUCHOS_FLOAT:INTERNAL=ON
HAVE_THYRA_FLOAT:INTERNAL=ON
HAVE_TPETRA_FLOAT:INTERNAL=ON
HAVE_TPETRA_FLOAT128:INTERNAL=OFF
HAVE_TPETRA_INST_FLOAT:INTERNAL=ON
HAVE_TPETRA_INST_FLOAT128:INTERNAL=OFF
MueLu_INST_FLOAT_INT_INT:INTERNAL=ON
Tpetra_INST_FLOAT_DEFAULT:INTERNAL=ON

$ grep _COMPLEX CMakeCache.txt
Anasazi_ENABLE_COMPLEX:BOOL=ON
Sacado_ENABLE_COMPLEX:BOOL=TRUE
Teuchos_ENABLE_COMPLEX:BOOL=TRUE
Thyra_ENABLE_COMPLEX:BOOL=ON
Tpetra_INST_COMPLEX_DOUBLE:BOOL=ON
Tpetra_INST_COMPLEX_FLOAT:BOOL=ON
Trilinos_ENABLE_COMPLEX:BOOL=ON
Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=ON
Trilinos_ENABLE_COMPLEX_FLOAT:BOOL=ON
CXX_COMPLEX_BLAS_WORKS:INTERNAL=1
CXX_COMPLEX_BLAS_WORKS_COMPILED:INTERNAL=TRUE
CXX_COMPLEX_BLAS_WORKS_EXITCODE:INTERNAL=0
HAVE_ANASAZI_COMPLEX:INTERNAL=ON
HAVE_COMPLEX_BLAS:INTERNAL=NOT;HAVE_COMPLEX_BLAS_PROBLEM;OR;HAVE_FIXABLE_COMPLEX_BLAS_PROBLEM;OR;HAVE_VECLIB_COMPLEX_BLAS
HAVE_COMPLEX_BLAS_PROBLEM:INTERNAL=FALSE
HAVE_KOKKOSTSQR_COMPLEX:INTERNAL=ON
HAVE_ML_COMPLEX_MAXWELL:INTERNAL=OFF
HAVE_MUELU_INST_COMPLEX_INT_INT:INTERNAL=ON
HAVE_SACADO_COMPLEX:INTERNAL=ON
HAVE_TEUCHOS_COMPLEX:INTERNAL=ON
HAVE_THYRA_COMPLEX:INTERNAL=ON
HAVE_TPETRA_COMPLEX_DOUBLE:INTERNAL=ON
HAVE_TPETRA_COMPLEX_FLOAT:INTERNAL=ON
HAVE_TPETRA_INST_COMPLEX_DOUBLE:INTERNAL=ON
HAVE_TPETRA_INST_COMPLEX_FLOAT:INTERNAL=ON
MueLu_INST_COMPLEX_INT_INT:INTERNAL=ON
Tpetra_INST_COMPLEX_DOUBLE_DEFAULT:INTERNAL=ON
Tpetra_INST_COMPLEX_FLOAT_DEFAULT:INTERNAL=ON
```

That all looks correct.

Now to try the build:

```
$ time make -j16 &> make.out

real    10m25.885s
user    121m28.055s
sys     7m13.338s
```

The build broke with:

```
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector.cpp:55:3:   required from here
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockView.hpp:348:5: error: no match for ‘operator!=’ in ‘alpha != 0.0’
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockView.hpp:348:5: note: candidates are:
In file included from /ascldap/users/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED/packages/tpetra/core/src/Tpetra_Map.hpp:1:0,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_DistObject_decl.hpp:53,
                 from /ascldap/users/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED/packages/tpetra/core/src/Tpetra_DistObject.hpp:1,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_MultiVector_decl.hpp:54,
                 from /ascldap/users/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED/packages/tpetra/core/src/Tpetra_MultiVector.hpp:1,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector_decl.hpp:45,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector.cpp:44:
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Map_decl.hpp:1593:6: note: template<class LocalOrdinal, class GlobalOrdinal, class Node> bool operator!=(const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>&, const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>&)
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Map_decl.hpp:1593:6: note:   template argument deduction/substitution failed:
In file included from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector_decl.hpp:46:0,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector.cpp:44:
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockView.hpp:348:5: note:   ‘const std::complex<float>’ is not derived from ‘const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>’
In file included from /home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.7.2/base/bin/../lib/gcc/x86_64-unknown-linux-gnu/4.7.2/../../../../include/c++/4.7.2/iterator:66:0,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/comm/src/Teuchos_DefaultMpiComm.hpp:66,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/comm/src/Teuchos_CommHelpers.hpp:58,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_ConfigDefs.hpp:97,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Map_decl.hpp:49,
                 from /ascldap/users/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED/packages/tpetra/core/src/Tpetra_Map.hpp:1,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_DistObject_decl.hpp:53,
                 from /ascldap/users/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED/packages/tpetra/core/src/Tpetra_DistObject.hpp:1,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_MultiVector_decl.hpp:54,
                 from /ascldap/users/rabartl/Trilinos.base/BUILDS/CHECKIN/MPI_RELEASE_DEBUG_SHARED/packages/tpetra/core/src/Tpetra_MultiVector.hpp:1,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector_decl.hpp:45,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector.cpp:44:
/home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.7.2/base/bin/../lib/gcc/x86_64-unknown-linux-gnu/4.7.2/../../../../include/c++/4.7.2/bits/stream_iterator.h:137:5: note: template<class _Tp, class _CharT, class _Traits, class _Dist> bool std::operator!=(const std::istream_iterator<_Tp, _CharT, _Traits, _Dist>&, const std::istream_iterator<_Tp, _CharT, _Traits, _Dist>&)
/home/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.7.2/base/bin/../lib/gcc/x86_64-unknown-linux-gnu/4.7.2/../../../../include/c++/4.7.2/bits/stream_iterator.h:137:5: note:   template argument deduction/substitution failed:
In file included from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector_decl.hpp:46:0,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector.cpp:44:
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_Experimental_BlockView.hpp:348:5: note:   ‘const std::complex<float>’ is not derived from ‘const std::istream_iterator<_Tp, _CharT, _Traits, _Dist>’
```

I noticed that float support in Tpetra was disabled by default and therefore my guess is that it has not been tested in a long time.  Therefore, one would expect this to be broken.

**4.2) `complex<double>` on:**

Configuring with only complex turned on (which should enabled `std::complex<double>` but not `std::complex<float>` since float is not enabled):

```
$ time ./do-configure -DTrilinos_ENABLE_COMPLEX=ON &> configure.out

real    2m37.708s
user    2m8.941s
sys     0m24.123s
```

The cmake configure output shows:

```
Processing explicit instantiation support for enabled packages ...

Processing ETI support: Ifpack2
-- Ifpack2: Processing ETI / test support
-- Enabled Scalar types:        double|std::complex<double>
-- Enabled LocalOrdinal types:  int
-- Enabled GlobalOrdinal types: int|long long
-- Enabled Node types:          Kokkos::Compat::KokkosSerialWrapperNode
-- User/Downstream ETI set: 
-- Excluded type combinations: 
-- Set of enabled types, before exclusions: S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long}
Processing ETI support: TpetraCore
-- TpetraCore: Processing ETI / test support
-- Enabled Scalar types:        int|long long|double|std::complex<double>
-- Enabled LocalOrdinal types:  int
-- Enabled GlobalOrdinal types: int|long long
-- Enabled Node types:          Kokkos::Compat::KokkosSerialWrapperNode
-- Excluded type combinations: :
-- Set of enabled types, before exclusions: S={int} LO={int} GO={int} N={Kokkos::Compat::KokkosSerialWrapperNode};S={int} LO={int} GO={long long} N={Kokkos::Compat::KokkosSerialWrapperNode};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={int} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={int} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={long long} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={long long} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={std::complex<double>} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long}
Processing ETI support: TpetraKernels
-- TpetraKernels: Processing ETI / test support
-- Enabled Scalar types:       int|long long|double|Kokkos::complex<double>
-- Enabled LocalOrdinal types: int
-- Enabled Device types:       Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>
-- Set of enabled types, before exclusions: S={int} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={long long} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={double} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={Kokkos::complex<double>} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>}
```

Looking at the CMakeCache.txt file:

```
$ grep _FLOAT CMakeCache.txt  | grep -vi complex
Teuchos_ENABLE_FLOAT:BOOL=FALSE
Thyra_ENABLE_FLOAT:BOOL=OFF
Tpetra_INST_FLOAT:BOOL=OFF
Tpetra_INST_FLOAT128:BOOL=OFF
Trilinos_ENABLE_FLOAT:BOOL=OFF
HAVE_MUELU_INST_FLOAT_INT_INT:INTERNAL=OFF
HAVE_TEUCHOS_FLOAT:INTERNAL=OFF
HAVE_THYRA_FLOAT:INTERNAL=OFF
HAVE_TPETRA_FLOAT:INTERNAL=OFF
HAVE_TPETRA_FLOAT128:INTERNAL=OFF
HAVE_TPETRA_INST_FLOAT:INTERNAL=OFF
HAVE_TPETRA_INST_FLOAT128:INTERNAL=OFF
MueLu_INST_FLOAT_INT_INT:INTERNAL=OFF
Tpetra_INST_FLOAT_DEFAULT:INTERNAL=OFF

$ grep _COMPLEX CMakeCache.txt
Anasazi_ENABLE_COMPLEX:BOOL=ON
Sacado_ENABLE_COMPLEX:BOOL=TRUE
Teuchos_ENABLE_COMPLEX:BOOL=TRUE
Thyra_ENABLE_COMPLEX:BOOL=ON
Tpetra_INST_COMPLEX_DOUBLE:BOOL=ON
Tpetra_INST_COMPLEX_FLOAT:BOOL=OFF
Trilinos_ENABLE_COMPLEX:BOOL=ON
Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=ON
Trilinos_ENABLE_COMPLEX_FLOAT:BOOL=OFF
CXX_COMPLEX_BLAS_WORKS:INTERNAL=1
CXX_COMPLEX_BLAS_WORKS_COMPILED:INTERNAL=TRUE
CXX_COMPLEX_BLAS_WORKS_EXITCODE:INTERNAL=0
HAVE_ANASAZI_COMPLEX:INTERNAL=ON
HAVE_COMPLEX_BLAS:INTERNAL=NOT;HAVE_COMPLEX_BLAS_PROBLEM;OR;HAVE_FIXABLE_COMPLEX_BLAS_PROBLEM;OR;HAVE_VECLIB_COMPLEX_BLAS
HAVE_COMPLEX_BLAS_PROBLEM:INTERNAL=FALSE
HAVE_KOKKOSTSQR_COMPLEX:INTERNAL=ON
HAVE_ML_COMPLEX_MAXWELL:INTERNAL=OFF
HAVE_MUELU_INST_COMPLEX_INT_INT:INTERNAL=ON
HAVE_SACADO_COMPLEX:INTERNAL=ON
HAVE_TEUCHOS_COMPLEX:INTERNAL=ON
HAVE_THYRA_COMPLEX:INTERNAL=ON
HAVE_TPETRA_COMPLEX_DOUBLE:INTERNAL=ON
HAVE_TPETRA_COMPLEX_FLOAT:INTERNAL=OFF
HAVE_TPETRA_INST_COMPLEX_DOUBLE:INTERNAL=ON
HAVE_TPETRA_INST_COMPLEX_FLOAT:INTERNAL=OFF
MueLu_INST_COMPLEX_INT_INT:INTERNAL=ON
Tpetra_INST_COMPLEX_DOUBLE_DEFAULT:INTERNAL=ON
Tpetra_INST_COMPLEX_FLOAT_DEFAULT:INTERNAL=OFF
```

That looks correct.

Now to try the build and testing:

```
$ time make -j16 &> make.out

real    75m23.951s
user    958m6.272s
sys     51m57.114s

$ time ctest -j16 &> ctest.out

real    11m57.753s
user    83m6.413s
sys     5m3.663s
```

That all built and the test summary is:

```
100% tests passed, 0 tests failed out of 2333

Label Time Summary:
Amesos               =  18.06 sec (13 tests)
Amesos2              =   9.35 sec (7 tests)
Anasazi              = 150.09 sec (100 tests)
AztecOO              =  17.39 sec (17 tests)
Belos                = 115.55 sec (74 tests)
Domi                 = 155.26 sec (125 tests)
Epetra               =  43.80 sec (61 tests)
EpetraExt            =  13.10 sec (10 tests)
FEI                  =  44.15 sec (43 tests)
Galeri               =   4.56 sec (9 tests)
GlobiPack            =   1.22 sec (6 tests)
Ifpack               =  61.23 sec (53 tests)
Ifpack2              =  48.53 sec (32 tests)
Intrepid             = 193.07 sec (152 tests)
Intrepid2            =  99.35 sec (107 tests)
Isorropia            =   7.67 sec (6 tests)
Kokkos               = 254.33 sec (21 tests)
ML                   =  44.97 sec (34 tests)
MueLu                = 317.75 sec (58 tests)
NOX                  = 150.07 sec (101 tests)
OptiPack             =   6.43 sec (5 tests)
Panzer               = 282.20 sec (125 tests)
Phalanx              =   2.72 sec (15 tests)
Pike                 =   1.75 sec (7 tests)
Piro                 =  26.13 sec (11 tests)
ROL                  = 664.45 sec (112 tests)
RTOp                 =  11.70 sec (24 tests)
Rythmos              = 165.63 sec (83 tests)
SEACAS               =   7.66 sec (8 tests)
STK                  =  12.96 sec (12 tests)
Sacado               =  51.88 sec (290 tests)
Shards               =   0.66 sec (4 tests)
ShyLU                =   8.30 sec (5 tests)
Stokhos              = 107.24 sec (74 tests)
Stratimikos          =  35.33 sec (39 tests)
Teko                 = 201.68 sec (19 tests)
Teuchos              =  37.78 sec (123 tests)
ThreadPool           =   7.75 sec (10 tests)
Thyra                =  61.53 sec (80 tests)
Tpetra               = 124.78 sec (122 tests)
TrilinosCouplings    =  62.56 sec (19 tests)
Triutils             =   2.23 sec (2 tests)
Xpetra               =  54.12 sec (16 tests)
Zoltan               = 198.06 sec (16 tests)
Zoltan2              = 125.15 sec (91 tests)
```

That all passed, which is not surprising because this is pretty close to the default configure setting for Trilinos prior to these changes.

Now to look for evidence of `std::complex<double>` tests and unit tests running ...

First note that the total number of tests run in 2333 which is 47 more than the build with no complex enabled which has 2286 tests.  That is explained partially by the addition of 44 more tests with "complex" in the name:

```
$ grep -i complex ctest.out | grep " Test " | wc -l
44

$ grep -i complex ctest.out | grep " Test "
479/2333 Test #1207: Belos_Tpetra_MVOPTester_complex_test_MPI_4 .........................................   Passed    1.82 sec
1183/2333 Test #2169: ROL_example_complex_example_01_MPI_1 ...............................................   Passed    1.16 sec
2288/2333 Test #1135: Belos_bl_cg_complex_hb_MPI_4 .......................................................   Passed    1.53 sec
2289/2333 Test #1136: Belos_bl_gmres_complex_hb_0_MPI_4 ..................................................   Passed    1.45 sec
2290/2333 Test #1137: Belos_bl_gmres_complex_hb_1_MPI_4 ..................................................   Passed    1.55 sec
2291/2333 Test #1142: Belos_MVOPComplexTester_MPI_4 ......................................................   Passed    1.31 sec
2292/2333 Test #1138: Belos_bl_gmres_complex_hb_2_MPI_4 ..................................................   Passed    2.02 sec
2293/2333 Test #1143: Belos_test_gcrodr_complex_MPI_4 ....................................................   Passed    1.49 sec
2294/2333 Test #1144: Belos_test_rcg_complex_MPI_4 .......................................................   Passed    1.22 sec
2295/2333 Test #1145: Belos_test_lsqr_complex_MPI_4 ......................................................   Passed    1.32 sec
2296/2333 Test #1146: Belos_test_factory_complex_MPI_4 ...................................................   Passed    1.32 sec
2297/2333 Test #1140: Belos_gcrodr_complex_hb_MPI_4 ......................................................   Passed    4.24 sec
2298/2333 Test #1148: Belos_pseudo_tfqmr_complex_hb_0_MPI_4 ..............................................   Passed    1.37 sec
2299/2333 Test #1147: Belos_tfqmr_complex_hb_MPI_4 .......................................................   Passed    1.38 sec
2300/2333 Test #1149: Belos_pseudo_tfqmr_complex_hb_1_MPI_4 ..............................................   Passed    1.88 sec
2301/2333 Test #1284: Anasazi_BlockDavidson_complex_lap_test_MPI_4 .......................................   Passed    1.26 sec
2302/2333 Test #1285: Anasazi_BlockDavidson_complex_test_0_MPI_4 .........................................   Passed    1.47 sec
2303/2333 Test #1286: Anasazi_BlockDavidson_complex_test_1_MPI_4 .........................................   Passed    1.30 sec
2304/2333 Test #1287: Anasazi_BlockDavidson_complex_test_2_MPI_4 .........................................   Passed    1.44 sec
2305/2333 Test #1288: Anasazi_BlockDavidson_complex_test_3_MPI_4 .........................................   Passed    1.44 sec
2306/2333 Test #1289: Anasazi_BlockKrylovSchur_complex_lap_test_MPI_4 ....................................   Passed    1.26 sec
2307/2333 Test #1290: Anasazi_BlockKrylovSchur_complex_test_0_MPI_4 ......................................   Passed    1.33 sec
2308/2333 Test #1294: Anasazi_LOBPCG_complex_lap_test_MPI_4 ..............................................   Passed    1.40 sec
2309/2333 Test #1293: Anasazi_BlockKrylovSchur_complex_test_3_MPI_4 ......................................   Passed    1.41 sec
2310/2333 Test #1292: Anasazi_BlockKrylovSchur_complex_test_2_MPI_4 ......................................   Passed    1.42 sec
2311/2333 Test #1291: Anasazi_BlockKrylovSchur_complex_test_1_MPI_4 ......................................   Passed    1.42 sec
2312/2333 Test #1296: Anasazi_IRTR_complex_lap_test_0_MPI_4 ..............................................   Passed    1.26 sec
2313/2333 Test #1298: Anasazi_IRTR_complex_lap_test_2_MPI_4 ..............................................   Passed    1.26 sec
2314/2333 Test #1295: Anasazi_LOBPCG_complex_test_MPI_4 ..................................................   Passed    1.47 sec
2315/2333 Test #1297: Anasazi_IRTR_complex_lap_test_1_MPI_4 ..............................................   Passed    1.26 sec
2316/2333 Test #1299: Anasazi_IRTR_complex_lap_test_3_MPI_4 ..............................................   Passed    1.31 sec
2317/2333 Test #1302: Anasazi_IRTR_complex_lap_test_6_MPI_4 ..............................................   Passed    1.27 sec
2318/2333 Test #1300: Anasazi_IRTR_complex_lap_test_4_MPI_4 ..............................................   Passed    1.28 sec
2319/2333 Test #1301: Anasazi_IRTR_complex_lap_test_5_MPI_4 ..............................................   Passed    1.27 sec
2320/2333 Test #1305: Anasazi_IRTR_complex_test_1_MPI_4 ..................................................   Passed    1.39 sec
2321/2333 Test #1304: Anasazi_IRTR_complex_test_0_MPI_4 ..................................................   Passed    1.39 sec
2322/2333 Test #1303: Anasazi_IRTR_complex_lap_test_7_MPI_4 ..............................................   Passed    1.39 sec
2323/2333 Test #1315: Anasazi_Tpetra_BlockDavidson_Complex_test_MPI_4 ....................................   Passed    1.53 sec
2324/2333 Test #1316: Anasazi_Tpetra_BlockKrylovSchur_ComplexLap_test_MPI_4 ..............................   Passed    1.54 sec
2325/2333 Test #1317: Anasazi_Tpetra_BlockKrylovSchur_Complex_test_0_MPI_4 ...............................   Passed    1.54 sec
2326/2333 Test #1318: Anasazi_Tpetra_BlockKrylovSchur_Complex_test_1_MPI_4 ...............................   Passed    1.68 sec
2327/2333 Test #1319: Anasazi_Tpetra_IRTR_ComplexLap_test_MPI_4 ..........................................   Passed    1.55 sec
2328/2333 Test #1320: Anasazi_Tpetra_IRTR_Complex_test_MPI_4 .............................................   Passed    1.54 sec
2329/2333 Test #1321: Anasazi_Tpetra_LOBPCG_Complex_test_MPI_4 ...........................................   Passed    1.54 sec
```

There should also be a bunch of unit tests with `std::complex<double>`:

```
$ grep -nH complex Testing/Temporary/LastTest.log | grep UnitTest | wc -l
1235
```

Make sure there are no `std::complex<float>` unit tests:

```
$ grep -nH complex Testing/Temporary/LastTest.log | grep float | wc -l
45
```

Looks like there are still some unit tests that are testing `std::complex<float>` even though they should not be.  But we can let that slide for now.

Otherwise, it looks like this build is working.  Later I will do a comparion of the runtime and storage costs for enabling `std::complex<double>` or now in the basic CI build in support of #482.  If it is not too high, then it might be okay to turn it on for pre-push CI testing but keep it disabled by default for builds of Trilinos.

**4.3) float on:**

Doing the configure:

```
$ time ./do-configure -DTrilinos_ENABLE_FLOAT=ON &> configure.out

real    2m36.837s
user    2m7.076s
sys     0m23.968s
```

The cmake configure output shows:

```
Processing explicit instantiation support for enabled packages ...

Processing ETI support: Ifpack2
-- Ifpack2: Processing ETI / test support
-- Enabled Scalar types:        float|double
-- Enabled LocalOrdinal types:  int
-- Enabled GlobalOrdinal types: int|long long
-- Enabled Node types:          Kokkos::Compat::KokkosSerialWrapperNode
-- User/Downstream ETI set: 
-- Excluded type combinations: 
-- Set of enabled types, before exclusions: S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long}
Processing ETI support: TpetraCore
-- TpetraCore: Processing ETI / test support
-- Enabled Scalar types:        int|long long|float|double
-- Enabled LocalOrdinal types:  int
-- Enabled GlobalOrdinal types: int|long long
-- Enabled Node types:          Kokkos::Compat::KokkosSerialWrapperNode
-- Excluded type combinations: :
-- Set of enabled types, before exclusions: S={int} LO={int} GO={int} N={Kokkos::Compat::KokkosSerialWrapperNode};S={int} LO={int} GO={long long} N={Kokkos::Compat::KokkosSerialWrapperNode};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={int} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={int} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={long long} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={long long} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={float} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={int};S={double} N={Kokkos::Compat::KokkosSerialWrapperNode} LO={int} GO={long long}
Processing ETI support: TpetraKernels
-- TpetraKernels: Processing ETI / test support
-- Enabled Scalar types:       int|long long|float|double
-- Enabled LocalOrdinal types: int
-- Enabled Device types:       Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>
-- Set of enabled types, before exclusions: S={int} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={long long} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={float} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>};S={double} LO={int} D={Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>}
```

The CMakeCache.txt files shows:

```
$ grep _FLOAT CMakeCache.txt  | grep -vi complex
Teuchos_ENABLE_FLOAT:BOOL=TRUE
Thyra_ENABLE_FLOAT:BOOL=ON
Tpetra_INST_FLOAT:BOOL=ON
Tpetra_INST_FLOAT128:BOOL=OFF
Trilinos_ENABLE_FLOAT:BOOL=ON
HAVE_MUELU_INST_FLOAT_INT_INT:INTERNAL=ON
HAVE_TEUCHOS_FLOAT:INTERNAL=ON
HAVE_THYRA_FLOAT:INTERNAL=ON
HAVE_TPETRA_FLOAT:INTERNAL=ON
HAVE_TPETRA_FLOAT128:INTERNAL=OFF
HAVE_TPETRA_INST_FLOAT:INTERNAL=ON
HAVE_TPETRA_INST_FLOAT128:INTERNAL=OFF
MueLu_INST_FLOAT_INT_INT:INTERNAL=ON
Tpetra_INST_FLOAT_DEFAULT:INTERNAL=ON

$ grep _COMPLEX CMakeCache.txt
Anasazi_ENABLE_COMPLEX:BOOL=OFF
Sacado_ENABLE_COMPLEX:BOOL=FALSE
Teuchos_ENABLE_COMPLEX:BOOL=FALSE
Thyra_ENABLE_COMPLEX:BOOL=OFF
Tpetra_INST_COMPLEX_DOUBLE:BOOL=OFF
Tpetra_INST_COMPLEX_FLOAT:BOOL=OFF
Trilinos_ENABLE_COMPLEX:BOOL=OFF
Trilinos_ENABLE_COMPLEX_DOUBLE:BOOL=OFF
Trilinos_ENABLE_COMPLEX_FLOAT:BOOL=OFF
HAVE_ANASAZI_COMPLEX:INTERNAL=OFF
HAVE_KOKKOSTSQR_COMPLEX:INTERNAL=OFF
HAVE_ML_COMPLEX_MAXWELL:INTERNAL=OFF
HAVE_MUELU_INST_COMPLEX_INT_INT:INTERNAL=OFF
HAVE_SACADO_COMPLEX:INTERNAL=OFF
HAVE_TEUCHOS_COMPLEX:INTERNAL=OFF
HAVE_THYRA_COMPLEX:INTERNAL=OFF
HAVE_TPETRA_COMPLEX_DOUBLE:INTERNAL=OFF
HAVE_TPETRA_COMPLEX_FLOAT:INTERNAL=OFF
HAVE_TPETRA_INST_COMPLEX_DOUBLE:INTERNAL=OFF
HAVE_TPETRA_INST_COMPLEX_FLOAT:INTERNAL=OFF
MueLu_INST_COMPLEX_INT_INT:INTERNAL=OFF
Tpetra_INST_COMPLEX_DOUBLE_DEFAULT:INTERNAL=OFF
Tpetra_INST_COMPLEX_FLOAT_DEFAULT:INTERNAL=OFF
```

That looks correct.

Now trying the build:

```
$ time make -j16 &> make.out

real    49m59.521s
user    630m52.833s
sys     32m53.624s
```

That failed showing the build failures:

```
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/muelu/test/unit_tests/Smoothers/Ifpack2Smoother.cpp:104:11: error: no matching function for call to ‘testRelErr(const char [14], Teuchos::ScalarTraits<float>::magnitudeType&, const char [22], double, const char [4], double, const char [4], double, Teuchos::Ptr<Teuchos::basic_FancyOStream<char> >)’
...
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/muelu/test/unit_tests/Smoothers/Ifpack2Smoother.cpp:104:11: error: in argument to unary !
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/muelu/test/unit_tests/Smoothers/Ifpack2Smoother.cpp: In instantiation of ‘void MueLuTests::Ifpack2Smoother_HardCodedResult_GaussSeidel2_UnitTest<Scalar, LocalOrdinal, GlobalOrdinal, Node>::runUnitTestImpl(Teuchos::FancyOStream&, bool&) const [with Scalar = float; LocalOrdinal = int; GlobalOrdinal = int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>; Teuchos::FancyOStream = Teuchos::basic_FancyOStream<char>]’:
...
```

Again, it looks like the float type has not been tested in a long time so it is not surpising that it is broken.

**4.5) float and `complex<double>` on:**

I am not going to bother testing this.  If this does not work, then who cares for now.

**5) Add documentation to enable of float and complex to Trilinos Build Reference and quick installation guide:**

[Done]

**Comment by bartlettroscoe** [comment]
@mhoemmen, I know this is just on a branch [ better-ci-build-482 ](https://github.com/bartlettroscoe/Trilinos/tree/better-ci-build-482), but can your please review the commit 89d74b3?  I adjusted the enable/disable for float and complex types in Tpetra and cleaned up the logic some.  You seem like the person who would most care to review this commit anyway.

**Comment by mhoemmen** [comment]
@bartlettroscoe I'm looking now.

**Comment by mhoemmen** [comment]
@bartlettroscoe Looks pretty sane to me.  I made some comments on the commit.  There's only one correctness issue I saw:  In anasazi/CMakeLists.txt, Tpetra_INST_COMPLEX_DOUBLE may not be defined, since it's possible to build Anasazi with Tpetra OFF.  One question: Should we add a Tpetra_INST_DOUBLE_DEFAULT, by analogy with Tpetra_INST_FLOAT_DEFAULT?

**Comment by mhoemmen** [comment]
I fixed the Tpetra build issues with Scalar = float and std::complex<float> (see #868).

**Comment by bartlettroscoe** [comment]
> There's only one correctness issue I saw: In anasazi/CMakeLists.txt, Tpetra_INST_COMPLEX_DOUBLE may not be defined, since it's possible to build Anasazi with Tpetra OFF

Okay, I updated the commit to separate out Teuchos and Tpetra complex checks in the commit 3f30267 that I just pushed to the branch [better-ci-build-482](https://github.com/bartlettroscoe/Trilinos/tree/better-ci-build-482).

> One question: Should we add a Tpetra_INST_DOUBLE_DEFAULT, by analogy with Tpetra_INST_FLOAT_DEFAULT

That is a good question.  Most other Trilinos packages don't even offer the ability to disable double support.  Tpetra seems to be the only package that allows one to disable double support.  It seems that Trilinos as a whole should decide if double support will always be built into the library or if it can be conditionally disabled.  I had assumed that nearly every customer would want support for double Scalars so there was no point in allowing it to be disabled (and then require us to set up builds that disable it to make sure things work with the different permutations of scalar types enabled).

I wanted to ask about why tpetra/CMakeLists.txt was exporting the defaults for many of these cache vars as global variables with `GLOBAL_SET()` statements?  The cache vars themselves like `Tpetra_INST_FLOAT` are global variables already.  Should not downstream packages just key off of those values and not the defaults?

Also, now that complex will be disabled by default, should the behavior of tpetra/CMakeLists.txt be changed so that when complex is explicitly enabled and HAVE_COMPLEX_BLAS is FALSE, then it errors out?  Currently, if the user enables Trilinos_ENABLE_COMPLEX_DOUBLE and HAVE_COMPLEX_BLAS is FALSE, then it silently skips support for complex in Tpetra since this sets the default for Tpetra_INST_COMPLEX_DOUBLE to FALSE.  Would it not be better if we set the default as:

```
IF (Trilinos_ENABLE_COMPLEX_DOUBLE)
  SET(Tpetra_INST_COMPLEX_DOUBLE_DEFAULT ON)
ELSE()
  SET(Tpetra_INST_COMPLEX_DOUBLE_DEFAULT OFF)
ENDIF()
```
 
and then raised an error if Tpetra_INST_COMPLEX_DOUBLE is ON and HAVE_COMPLEX_BLAS is OFF?  It seems that if a user explicitly turns on complex and they don't get it, then that should be an error.  Would not not be less surprising for users than the current behavior?

**Comment by mhoemmen** [comment]
> That is a good question. Most other Trilinos packages don't even offer the ability to disable double support. Tpetra seems to be the only package that allows one to disable double support. It seems that Trilinos as a whole should decide if double support will always be built into the library or if it can be conditionally disabled. I had assumed that nearly every customer would want support for double Scalars so there was no point in allowing it to be disabled (and then require us to set up builds that disable it to make sure things work with the different permutations of scalar types enabled).

That's reasonable.  Nobody has ever asked me to disable Scalar=double ;-)

> I wanted to ask about why tpetra/CMakeLists.txt was exporting the defaults for many of these cache vars as global variables with GLOBAL_SET() statements? The cache vars themselves like Tpetra_INST_FLOAT are global variables already. Should not downstream packages just key off of those values and not the defaults?

Use of GLOBAL_SET was likely just an artifact of different people working on that file, copying and pasting CMake code and hoping it would work ;-)  I don't think we need to export those defaults.

> Also, now that complex will be disabled by default, should the behavior of tpetra/CMakeLists.txt be changed so that when complex is explicitly enabled and HAVE_COMPLEX_BLAS is FALSE, then it errors out?

I think that's a good idea.

Could you please file issues for the two suggested changes you mentioned?  I think those are good ideas.

**Comment by bartlettroscoe** [comment]
> Could you please file issues for the two suggested changes you mentioned? I think those are good ideas.

I will file them after I merge and push the `better-ci-build-482` branch.  That way you will not get a merge conflict when making these changes.  Thanks for the review!

**Comment by bartlettroscoe** [comment]
This has been merged to Trilinos 'develop' branch for a long time.  In am not going to bother with the task 

* Explicitly turn on Trilinos_ENABLE_FLOAT, Trilinos_ENABLE_COMPLEX, Trilinos_ENABLE_COMPLEX_FLOAT and Trilinos_ENABLE_COMPLEX_DOUBLE in a few automated Trilinos builds.

Closing as complete.
